### DISCUSSION TITLE
Are user enrolled source functions able to update the ruser_mesh_data arrays?

### DISCUSSION THREAD

**Discussion by hwhitehead** [PrincetonUniversity/athena#464] [Q&A] [answered]
I am attempting to implement a source function that removes mass from gas in the simulation and add it to a central body. The mass of this body is tracked in one the ruser_mesh_data Athena arrays. I am able to read data from this array by backtracking form the meshblock pointer, up to the mesh and then down to ruser_mesh_data array like this:

Real m = pmb->pmy_mesh->ruser_mesh_data(0)(1); // here I access the 2nd value in the 1st ruser array

However I also want to then change the value in this array (to account for mass gained during the timestep). The following line does not seem to update the array.

pmb->pmy_mesh->ruser_mesh_data(0)(1) = 10; // here I attempt to change the 2nd value in the 1st array to 10

I am confused as to why this does not work, as the user defined functions are passed the MeshBlock pointer and can therefore 'see' the array in the parent mesh. 

Is this an issue anyone else has come across?

**Comment by c-white** [comment]
This should work, but here are a couple thoughts on what might be going wrong:

1. The first index should be in brackets: `ruser_mesh_data[0](1)`.
2. Are the arrays properly allocated? `ruser_mesh_data` itself needs to be allocated via `AllocateRealUserMeshDataField`, and then `ruser_mesh_data[0]` needs to be allocated via `NewAthenaArray`. Both of these should probably happen in `InitUserMeshData`.
3. Issues with parallelization: There is one collection of `ruser_mesh_data` arrays per Mesh (i.e. per MPI rank). This means that if you have multiple MeshBlocks per rank, they will be reading and writing to the same place. Also, if you have multiple MPI ranks, they will not see each other's `ruser_mesh_data` without explicitly writing new MPI communication.

**Reply by hwhitehead** [reply]
1. in the actual code, the square brackets are there, they just changed format when typed in the post (I'm new to the forums) 
2. I believe the arrays are properly allocated, I have been following the wiki and have completed each of the steps 
3. This code here was running in serial (I have had separate issues with parallelisation), but ideally I'd like this data to be shared across ranks. Do you have any advice for passing information between MPI ranks within Athena++?

I thought ruser_mesh_data would be shared globally, but if it is rank specific it might not be the right tool. Do you know if there any variable spaces that are global viewable and accessible by each of the Meshes?

It remains somewhat peculiar that this does not work in serial, as each MeshBlock should only be writing to the mesh one at a time. I wonder what might be preventing it from doing it in this case.

**Reply by c-white** [reply]
I think the details of what's going wrong will require the context of the rest of the code. Can you post the pgen file?

As for simulation-scope variables, the code doesn't have any built-in way to do this, for two reasons. First the beauty of hyperbolic PDEs is that their evolution is local; a cell only needs to know about its immediate neighbors to advance in time. By making a dependence on a global variable, the problem is turning into something integro-differential. Second, this requires an all-to-all MPI communication of the sort that wasn't needed before, and it's not clear where in the algorithm this needs to be done; different needs will require different communication solutions. The new communication could also prove very expensive.

For these reasons, my first piece of advice to people wanting to make a global variable is to consider whether such a thing is truly needed.

If the answer is yes -- and sometimes it is -- then you'll probably want to add something like this somewhere in the code.

```
#ifdef MPI_PARALLEL
{
  MPI_Allreduce(MPI_IN_PLACE, ruser_mesh_data[0].data(), ruser_mesh_data[0].GetDim1(),
      MPI_ATHENA_REAL, MPI_SUM, MPI_COMM_WORLD);
}
#endif
```

Maybe instead of a sum-reduction you'll want some other operation. And where exactly this goes depends on where the variables are being read and where they are being modified in your code. Most importantly, this code needs to be guarded against the possibility of having more than one MeshBlock per Mesh, when you want this called once per Mesh (i.e. rank). Especially tricky is the possibility of having unequal numbers of MeshBlocks per Mesh, in which case unprotected code like this will hang the simulation.

**Reply by hwhitehead** [reply]
sent my response as a separate comment accidentally, see below.

**Comment by hwhitehead** [comment]
Here is a link to the pgen file: [link](https://github.com/hwhitehead/sharedfiles/blob/main/flow.cpp). The setup is a central potential gravitationally attracting gas in the shearing box prescription. What I want to do is update the mass of the central body as gas is removed from the simulation. I believe this is necessary as Athena++ does not have a prescription for particles (to my knowledge). 

The code is adapted from the ssheet pgen and so has a lot of baggage. I'll point to the key parts:
- allocation of the ruser_mesh_data in InitUserMeshData 
- enrolment of source function 'sink' in InitUserMeshData (as well as declaration at file start)
- the full sink function, which in its last line attempts to change the value of an array member of ruser_mesh_data
- InitUserMeshBlockData, which allocates an output variable
- UserWorkBeforeOutput that takes the (supposedly) updated value form ruser_mesh_data and adds it to the output (for analysis)

With the code in its current configuration, the 'mass' value does not appear to be changed by the sink function. This happens when the code is run as a single block in serial. 

I hope you can see why I think I need a global variable to handle this. One solution might be to store the mass changes at the block level and then introduce a routine to collect the block data at the end of the time step to the mesh. I would then need a routine to share the data across the mpi ranks (between meshes). 

Thanks so much for looking into this, I really appreciate the help.

**Comment by c-white** [comment, marked as answer]
I think I see the immediate problem. `EnrollUserExplicitSourceFunction` doesn't add to a list of source functions; it only replaces the one source function allowed. In this case, the thing to do is enroll a simple source function -- call it `AllSources` or something -- and have that function make calls to your other functions as needed.

**Reply by hwhitehead** [reply]
Ah thank you so much, I'm surprised I hadn't noticed the sources overwriting each other. I'll experiment more with this MPI mesh passing problem next week, Marking as solved!
